### PR TITLE
main: use cobra subcommands, implement list command

### PR DESCRIPTION
**Pull request by winny-**

### CODE REVIEW

File: nvc/extract.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+	return e, nil
+}
+
+func readCount(r io.Reader) (uint32, error) {
+	var count uint32
+	if err := binary.Read(r, binary.LittleEndian, &count); err != nil {
+		return 0, err
+	}
+	return count, nil
+}
```
Comment (sector-f) [inline nvc/extract.go:69]: Is there a reason to have these (and `readNVCMagic`) as separate functions? It might be better to have a single function that takes an `io.Reader` and returns a `([]nvcEntry, error)`. I think that would also reduce code duplication between the two subcommands.

Also, consistency in capitalization in function names (NVC vs Nvc) would be nice (but I really don't care which one we go with as long as we keep it consistent)

File: nvc/extract.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	Hash        uint64
+	Offset      uint32
+	RawLength   uint32 // Uncompressed length
+	Length      uint32 // Actual length in file
+	Compression compressionType
+}
+
+func Hash2String(hash uint64) string {
+	return fmt.Sprintf("%016x", hash)
+}
```
Comment (sector-f) [inline nvc/extract.go:34]: Maybe we could do `type hash uint64` and then rename this function `func (h hash) String()` (to implement the `Stringer` interface)? Not really important though

File: cmd/list.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+	Long: `based off jh_extract.py`,
+	Args: cobra.ExactArgs(1),
+	Run: func(cmd *cobra.Command, args []string) {
+		reader, openErr := os.Open(args[0])
+		if (openErr != nil) {
+			fmt.Println(openErr)
+			os.Exit(1)
+		}
+		entries, listErr := nvc.ListEntries(reader)
+		if (listErr != nil) {
```
Comment (sector-f) [inline cmd/list.go:22]: No parentheses needed in boolean expressions

File: cmd/list.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+)
+
+var listCmd = &cobra.Command{
+	Use:   "list",
+	Short: "Manipulate nvc files",
+	Long: `based off jh_extract.py`,
+	Args: cobra.ExactArgs(1),
+	Run: func(cmd *cobra.Command, args []string) {
+		reader, openErr := os.Open(args[0])
+		if (openErr != nil) {
```
Comment (sector-f) [inline cmd/list.go:17]: No parentheses needed in boolean expressions

File: cmd/list.go [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+			fmt.Println(openErr)
+			os.Exit(1)
+		}
+		entries, listErr := nvc.ListEntries(reader)
+		if (listErr != nil) {
+			fmt.Println(listErr)
+			os.Exit(1)
+		}
+		for _, entry := range(entries) {
```
Comment (sector-f) [inline cmd/list.go:26]: This works? If so, I honestly didn't even know you could do that

`range entries`, not `range(entries)`

File: cmd/list.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+	Args: cobra.ExactArgs(1),
+	Run: func(cmd *cobra.Command, args []string) {
+		reader, openErr := os.Open(args[0])
+		if (openErr != nil) {
+			fmt.Println(openErr)
+			os.Exit(1)
+		}
+		entries, listErr := nvc.ListEntries(reader)
+		if (listErr != nil) {
+			fmt.Println(listErr)
```
Comment (sector-f) [inline cmd/list.go:23]: Nit: printing errors to stderr is preferable, e.g. `fmt.Fprintln(os.Stderr, listErr)`

File: cmd/list.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
+package cmd
+
+import 	(
+	"fmt"
+	"os"
+	"github.com/spf13/cobra"
```
Comment (sector-f) [inline cmd/list.go:6]: Put an empty line between these lines to separate the stdlib imports from external imports

File: cmd/root.go [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+	Use:   "jh_extract",
+	Short: "Manipulate nvc files",
+	Long: `based off jh_extract.py`,
+	Run: func(cmd *cobra.Command, args []string) {
+		fmt.Println("durr")
+	},
+}
+
+func init() {
+	fmt.Println("in root")
```
Comment (sector-f) [inline cmd/root.go:19]: I'm guessing this was just for testing `init`'s behavior and can be removed now

File: cmd/root.go [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+	"github.com/spf13/cobra"
+)
+
+var rootCmd = &cobra.Command{
+	Use:   "jh_extract",
+	Short: "Manipulate nvc files",
+	Long: `based off jh_extract.py`,
+	Run: func(cmd *cobra.Command, args []string) {
+		fmt.Println("durr")
+	},
```
Comment (sector-f) [inline cmd/root.go:15]: You can just omit `Run` from the root function; I'm pretty sure it'll print a help message if no subcommand is given in that case

File: nvc/listFiles.go [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+	entries := make([]nvcEntry, count)
+
+	for i := range(entries) {
+		entry, eErr := readNvcEntry(r)
+		if eErr != nil {
+			return nil, eErr
+		}
+		entries[i] = entry
+	}
```
[the remark this answers is not in the document]
Comment (winny-) [reply nvc/listFiles.go:25]: From a testing perspective, would be easier to test interfaces that operate on `io.Reader`?  This could also allow for composing functions as an easy way to handle edge cases of the format, such as format changes etc.

[the remark this answers is not in the document]
Comment (sector-f) [reply nvc/listFiles.go:25]: Maybe keep the existing separate functions but add an additional helper function that just calls all of them in the correct order?

File: nvc/extract.go (hunk 1 above)
Comment (winny-) [reply nvc/extract.go:69]: Since it's a separate package, I'm curious if we could remove the NVC from most of the constants/function names?

I've also gone ahead and combined the readCount and readEntry into one public function, since they probaly are always used together.

File: nvc/extract.go (hunk 2 above)
Comment (winny-) [reply nvc/extract.go:34]: I'll add a TODO.

File: nvc/extract.go (hunk 1 above)
Comment (sector-f) [reply nvc/extract.go:69]: > Since it's a separate package, I'm curious if we could remove the NVC from most of the constants/function names?

Yeah, good point. `nvc.Entry` does read better than `nvc.nvcEntry`. There's probably something in Effective Go (or something) about that that.

> I've also gone ahead and combined the readCount and readEntry into one public function, since they probaly are always used together.

Right, and at that point, if you need the number of entries.....`len(entries)` :stuck_out_tongue:

Comment (sector-f) [reply nvc/extract.go:69]: Yep, found it: https://go.dev/doc/effective_go#package-names

> The importer of a package will use the name to refer to its contents, so exported names in the package can use that fact to avoid repetition...For instance, the buffered reader type in the bufio package is called Reader, not BufReader, because users see it as bufio.Reader, which is a clear, concise name.

File: cmd/list.go (hunk 3 above)
Comment (winny-) [reply cmd/list.go:22]: :boom:

File: cmd/list.go (hunk 4 above)
Comment (winny-) [reply cmd/list.go:17]: :boom:

File: cmd/list.go (hunk 5 above)
Comment (winny-) [reply cmd/list.go:26]: :boom:

File: cmd/list.go (hunk 6 above)
Comment (winny-) [reply cmd/list.go:23]: :boom:

File: cmd/root.go (hunk 9 above)
Comment (winny-) [reply cmd/root.go:15]: :boom:

File: nvc/extract.go [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+)
+
+func ReadMagic(r io.Reader) error {
```
Comment (sector-f) [inline nvc/extract.go:11]: This function doesn't need to be exported (does it?)

File: nvc/types.go [hunk 12]
```diff
@@ -0,0 +1,19 @@
+package nvc
+
+type EntryFlags uint32
+
+const (
+	UNCOMPRESSED EntryFlags = 0
+	COMPRESSED   EntryFlags = 1
+	ENCRYPTED    EntryFlags = 2
+
+	MAGIC = "nvc1d\x00\x00\x00"
```
Comment (sector-f) [inline nvc/types.go:10]: These names should just be lowercase  
(remember that identifiers that start with an uppercase letter are exported; these don't need to be)

**Comment by winny-** [comment]
```
[nix-shell:~/jh/jh_extract]$ ./jh_extract list ../core.nvc  | awk -F' |=' '{ print $9, $1; }' | sort | uniq --check-chars=1
1 001607f90729810a
3 1bac1592d74458d8

[nix-shell:~/jh/jh_extract]$ ./jh_extract list ../assets.nvc  | awk -F' |=' '{ print $9, $1; }' | sort | uniq --check-chars=1
0 00023ae48d345f9f
1 00211cd0c8a4dbbb
```

flags found in each file

File: cmd/list.go [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+			fmt.Fprintln(os.Stderr, listErr)
+			os.Exit(1)
+		}
+		for _, entry := range entries {
+			fmt.Printf("%v offset=%v %vB (%vB on disk) flags=%v\n",
+				nvc.Hash2String(entry.Hash),
+				entry.Offset,
+				entry.RawLength,
+				entry.Length,
+				entry.Flags)
```
Comment (sector-f) [inline cmd/list.go:32]: Might as well make `nvc.TocEntry` implement the `Stringer` interface at this point?

File: nvc/listFiles.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (winny-) [reply nvc/listFiles.go]: good idea, we should work towards that

File: nvc/toc.go [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+//     a. FNV1a hash (64-bit LE unsigned integer)
+//     b. Offset in the NVC archive (64-bit LE unsigned integer)
+//     c. Uncompressed length of the Entry (64-bit LE unsigned integer)
+//     d. Actual length of the Entry (64-bit LE unsigned integer)
+//     e. Entry flags (64-bit LE unsigned integer)
+func ReadToc(r io.Reader) ([]TocEntry, error) {
+	count, countErr := readCount(r)
+	if countErr != nil {
+		return nil, countErr
+	}
```
Comment (sector-f) [inline nvc/toc.go:21]: I think checking the magic bytes should be moved to the start of this function (so that we can then remove that check from everywhere else).

Right now, `ListEntries` is just "check for magic bytes then call this function".  `extractNVC` also has some code that....checks the magic bytes and then calls this function. I don't think any caller that just wants a list of ToC entries should have to worry about something as low-level as magic bytes.

To be clear, the change I'm proposing would include deleting the `ListEntries` function.

File: nvc/extract.go [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+	}
+
+	if string(header[:]) != MAGIC {
+		return errors.New(".nvc signature not found")
+	}
+
+	return nil
+}
+
+func extractNVC(arcPath string, pathlist []string, outputDirectory string, extractUnknown bool) error {
```
Comment (sector-f) [inline nvc/extract.go:24]: I think we're approaching the point where this function should be moved out of the `nvc` package (and into a cobra subcommand). This function is really application-level code; I don't think it belongs in this package.

That should be handled in a separate PR though.

File: nvc/toc.go (hunk 14 above)
Comment (winny-) [reply nvc/toc.go]: :boom:

File: cmd/list.go (hunk 13 above)
Comment (winny-) [reply cmd/list.go]: :boom:

File: nvc/extract.go (hunk 11 above)
Comment (winny-) [reply nvc/extract.go]: :boom:

File: nvc/types.go (hunk 12 above)
Comment (winny-) [reply nvc/types.go:11]: i saw some conficting style recommendations, what is the standard ?

File: nvc/extract.go (hunk 15 above)
Comment (winny-) [reply nvc/extract.go]: :boom:

File: cmd/extract.go [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+	}
+
+	hashToPath := map[nvc.Hash]string{}
+	for _, p := range pathlist {
+		hash := nvc.String2Hash(p)
+		hashToPath[hash] = p
+	}
+
+	var tocEntries []nvc.TocEntry
+	tocEntries, err = nvc.ReadToc(r)
```
Comment (sector-f) [inline cmd/extract.go:24]: Just do `tocEntries, err := nvc.ReadToc(r)` (instead of declaring `tocEntries` separately)

File: nvc/hash.go [hunk 17]
```diff
@@ -0,0 +1,20 @@
+package nvc
+
+import (
+	"fmt"
+	"hash/fnv"
+)
+
+func Hash2String(hash Hash) string {
+	return fmt.Sprintf("%016x", uint64(hash))
```
Comment (sector-f) [inline nvc/hash.go:9]: Move this to the `String` method and delete the `Hash2String` function

File: nvc/toc.go [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+		return nil, magicErr
+	}
+	count, countErr := readCount(r)
+	if countErr != nil {
+		return nil, countErr
+	}
+
+	entries := make([]TocEntry, count)
+
+	for i := range(entries) {
```
Comment (sector-f) [inline nvc/toc.go:31]: Nit: I think it would be better to do `for i:=0; i<count; i++` here rather than ranging over `entries`.

File: cmd/extract.go [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+
+	for hash, nvcEntry := range entries {
+		switch nvcEntry.Flags {
+		case nvc.Uncompressed:
+			// Do nothing
+		case nvc.Compressed:
+			// TODO: Perform zlib decompression
+		default:
+			continue
+		}
```
Comment (sector-f) [inline cmd/extract.go:44]: This is the wrong approach if we're going to treat these as bitmask flags (i.e. encrypted is 2 rather than 3).

Should probably have something like

```go
var (
    compressed bool
    encrypted  bool
)
```

And then do some bit-twiddling to assign the correct values to those variables (and then act accordingly)

File: cmd/extract.go (hunk 16 above)
Comment (winny-) [reply cmd/extract.go]: :boom:

File: nvc/hash.go (hunk 17 above)
Comment (winny-) [reply nvc/hash.go]: :boom:

File: cmd/extract.go (hunk 19 above)
Comment (sector-f) [reply cmd/extract.go:43]: Can we change the value of `nvc.Encrypted` back to 3 for now then?

File: cmd/list.go (hunk 7 above)
Comment (sector-f) [reply cmd/list.go:6]: This still needs to be done

File: cmd/root.go (hunk 8 above)
Comment (sector-f) [reply cmd/root.go]: This still needs to be done

Comment (sector-f) [reply cmd/root.go]: This still needs to be done

File: cmd/root.go (hunk 9 above)
Comment (sector-f) [reply cmd/root.go]: This still needs to be done

File: cmd/extract.go (hunk 19 above)
Comment (winny-) [reply cmd/extract.go:43]: woops fixed i

File: cmd/list.go (hunk 7 above)
Comment (winny-) [reply cmd/list.go]: woops fixed it
